### ISSUE TITLE
Interactive Support only for SVG Renderer?

### ISSUE THREAD

**Issue by JinCaoTH** [jonobr1/two.js#306]
I am sorry if the question is trivial.  In the docs, I only see one example featuring mouse event response through svg element. Does two.js only have interactive support for svg renderer?  Thanks!

**Comment by jonobr1** [comment]
That is correct. There isn't hit-test logic built into Two.js yet. So, at the moment the only manner I'd through SVG.

**Comment by jonobr1** [comment]
This ticket will serve as a good thread to start tracking features on hit test logic.

**Comment by jonobr1** [comment]
Thanks for sharing. That's a great point of reference and looks like it could be a good avenue to implement not just hit testing but frustum culling as well.

**Comment by jonobr1** [comment]
## Proposed Solution: Lightweight Bounds-Based Hit Detection

After researching the codebase and analyzing several approaches, I recommend implementing **lightweight bounds-based hit detection** that works across all renderers (SVG, Canvas, WebGL).

### Why This Approach?

**1. Aligns with Library Philosophy**
- Two.js is a *rendering* library, not an interaction framework
- The codebase is lean and focused - adding minimal, well-scoped features fits the existing architecture
- Purely additive API that doesn't break existing code

**2. Solves 80% of Use Cases**
- Most applications have <100 interactive shapes
- Simple AABB (axis-aligned bounding box) checks are fast enough for typical scenarios
- Complex apps can implement custom spatial indexing if needed

**3. Lowest Maintenance Burden**
- No external dependencies to track
- Small, testable surface area
- Easy to document and support
- Leverages existing `getBoundingClientRect()` infrastructure

**4. Fast Time-to-Value**
- Can be implemented in 2-3 weeks
- Users get cross-renderer interaction support quickly
- Can be enhanced later with spatial indexing if demand exists

**5. Backwards Compatible**
- Purely additive API
- Doesn't break existing code
- SVG DOM event approach still works

### Proposed API

```javascript
// Check if a point is inside a shape
const circle = two.makeCircle(100, 100, 50);
circle.contains(120, 110); // true

// Find shapes at a point
const shapes = two.getShapesAtPoint(mouseX, mouseY);
if (shapes.length > 0) {
  shapes[0].fill = 'red';
}

// Works with all renderers and transformations
const rect = two.makeRectangle(200, 200, 100, 50);
rect.rotation = Math.PI / 4;
rect.contains(220, 210); // handles rotation correctly
```

### Implementation Plan

**Phase 1: Core hit detection methods** (Week 1)
- Add `contains(x, y)` to `Path`, `Circle`, `Rectangle`, `Ellipse`, `Line`, `Polygon`
- Use AABB for fast rejection, then precise geometric tests
- Support transformations (rotation, scale, translation)
- Add comprehensive test coverage

**Phase 2: Scene utilities** (Week 2)
- Add `Two.getShapesAtPoint(x, y, options)` to traverse scene graph
- Support options: `visibleOnly`, `deepest` (only top shape), `all` (all intersecting shapes)
- Add coordinate transformation helpers for screen-to-world conversion
- Handle group hierarchies correctly

**Phase 3: Documentation & examples** (Week 3)
- Interactive shapes example showing click detection
- Draggable objects example
- Hover effects example
- Examples for all three renderers (SVG, Canvas, WebGL)
- Update API documentation

### Implementation Details

Each shape would implement `contains(x, y)` using:

- **Circle**: Distance test from center
- **Rectangle**: Rotated bounding box test  
- **Ellipse**: Ellipse equation test
- **Path**: AABB culling first, then optional precise point-in-polygon (ray casting)
- **Line**: Distance from point to line segment
- **Polygon**: Point-in-polygon test

All tests account for world transformations by using the existing `worldMatrix`.

### Why Not More Complex Solutions?

I also considered:
- **Full event system with spatial indexing** (rbush) - Over-engineered for library scope, adds external dependencies
- **Plugin architecture** - Premature optimization, better to start simple
- **Renderer-specific implementations** - Violates DRY, inconsistent developer experience

These could be explored later if demand exists, but starting simple provides the best balance of developer experience, performance, and maintainability.

### Success Metrics

After implementation:
- ✅ Enable click/hover detection on Canvas and WebGL (not just SVG)
- ✅ Work correctly with transformed shapes (rotation, scale, groups)
- ✅ Perform well with typical scene sizes (100-200 shapes)
- ✅ Add minimal bundle size impact (<2KB minified)
- ✅ Maintain 100% backwards compatibility

I'm happy to work on a PR implementing this approach if there's interest!

**Comment by jonobr1** [comment]
Addressed with: https://github.com/jonobr1/two.js/pull/801
